build-info-extractor-pip/src/main/java/org/jfrog/build/extractor/pip/extractor/PipBuildInfoExtractor.java

```diff
@@ -57,58 +59,126 @@ Build extract(ArtifactoryDependenciesClient client, String repository, String in
-                missingDeps.add(pkgName);
+                // Dependency downloaded from Artifactory.
+                getFromArtifactoryMap.put(fileName, pkgName);
                 continue;
             }
+            // Dependency wasn't downloaded in this execution, get dependency info from cache.
+            if (dependenciesCache != null && dependenciesCache.getDependency(pkgName) != null) {
+                dependenciesMap.put(pkgName, dependenciesCache.getDependency(pkgName));
+            }
+        }
 
-            dependenciesMap.put(pkgName, dependency);
+        // Get dependencies from Artifactory.
+        Map<String, Dependency> dependenciesFromArtifactory = getDependenciesFromArtifactory(getFromArtifactoryMap, repository, client, logger);
+        if (dependenciesFromArtifactory != null) {
+            dependenciesMap.putAll(dependenciesFromArtifactory);
         }
 
+        // Prompt missing dependencies.
+        Set<String> missingDeps = downloadedDependencies.keySet().stream()
+                .filter(x -> !dependenciesMap.containsKey(x))
+                .collect(Collectors.toSet());
         promptMissingDeps(missingDeps, logger);
+
         DependenciesCache.updateDependenciesCache(dependenciesMap, executionPath);
         return dependenciesMap;
     }
 
-    Dependency getPackageInfoFromArtifactory(String fileName, String repository, ArtifactoryDependenciesClient client, Log logger) throws IOException {
-        String aql = String.format(PIP_AQL_FORMAT, repository, fileName);
-        AqlSearchResult searchResult;
+    /**
+     * Fetch required files-data from Artifactory and create its corresponding Dependency object.
+     * @param fileToPackageMap - Mapping between a downloaded file to its package name.
+     * @param repository - Resolution repository.
+     * @param client - Artifactory client for fetching artifacts data.
+     * @param logger - The logger.
+     * @return Mapping of a package-name and its Dependency object.
+     * @throws IOException
+     */
+    private Map<String, Dependency> getDependenciesFromArtifactory(Map<String, String> fileToPackageMap, String repository,
+                                                                   ArtifactoryDependenciesClient client, Log logger) throws IOException {
+        if (fileToPackageMap.isEmpty()) {
+            return null;
+        }
+
+        // Create queries.
+        List<String> aqlQueries = new ArrayList<>();
+        StringBuilder filesQueryPartBuilder = new StringBuilder();
+        int filesCounter = 0;
+        for (String file : fileToPackageMap.keySet()) {
+            filesCounter++;
+            filesQueryPartBuilder.append(String.format(PIP_AQL_FILE_PART, file));
+            if (filesCounter == PIP_AQL_BULK_SIZE) {
+                aqlQueries.add(getPipDependenciesAql(filesQueryPartBuilder, repository));
+                filesCounter = 0;
+                filesQueryPartBuilder.setLength(0);
```

Comment:
Better to create a new StringBuilder here.

Reply:
Can you please explain? For my knowledge it may cause much overhead to create a new one instead of using the previous after setting length to 0.

Reply:
This is not critical. You can leave it as is. I don't think there's any overhead either way. 
```new StringBuilder()``` is a bit more "conventional" than ```filesQueryPartBuilder.setLength(0)```.